06_puppy/hsy3418/mapStore.go

```diff
@@ -0,0 +1,50 @@
+package main
+
+import "fmt"
+
+// MapStore is a implementation of storer for the storage of puppies
+type MapStore struct {
+	puppies map[int32]Puppy
+	nextID  int32
+}
+
+// NewMapStore creates a new mapStore
+func NewMapStore() *MapStore {
+	return &MapStore{
+		puppies: map[int32]Puppy{},
+	}
+}
+
+// CreatePuppy adds a nuw puppy to the puppy store and returns the id for the puppy
+func (m *MapStore) CreatePuppy(puppy Puppy) int32 {
+	puppy.ID = m.nextID
+	m.nextID++
+	m.puppies[puppy.ID] = puppy
+	return puppy.ID
+}
+
+// ReadPuppy retrieves the puppy for a given id from puppies store
+func (m *MapStore) ReadPuppy(id int32) (Puppy, error) {
+	if _, ok := m.puppies[id]; !ok {
+		return Puppy{}, fmt.Errorf("puppy with %d ID does not exist", id)
+	}
+	return m.puppies[id], nil
```

Comment:
Instead of looking up the map again, when you did it already before, you should save the value from from the first lookup. It means you can't put it all inside the `if`, but that doesn't matter:
```
puppy, ok := m.puppies[id]
if !ok {
    ...
}
return puppy, nil
```